watcher.go

```diff
@@ -0,0 +1,141 @@
+	// This is copied from DefaultTransport.
+	// we don't do an actual copy because the Trasnport uses mutexes.
+	// copied mutexes are ineffective.
+	// You will panic on concurrent map access if running
+	// with other clients without instantiating a new transport here.
+	ts := &http.Transport{
+		DialContext: (&net.Dialer{
+			Timeout: 5 * time.Second,
+		}).DialContext,
+		DisableCompression:    true,
+		MaxIdleConns:          5,
+		MaxIdleConnsPerHost:   2,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   5 * time.Second,
+		ResponseHeaderTimeout: 10 * time.Minute,
+		ExpectContinueTimeout: 5 * time.Second,
+	}
+	c := &Client{
+		Transport: ts,
+	}
+	return c
+}
+
+// Watch is the package-level Watch defintion which is called on
+// DefaultWatcher.
+func Watch(ctx context.Context, key string, handler WatcherFunc) {
+	DefaultWatcher.Watch(ctx, key, handler)
+}
+
+// WatchPrefix is the package-level WatchPrefix defintion which is called on
+// DefaultWatcher.
+func WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	DefaultWatcher.WatchPrefix(ctx, prefix, handler)
+}
+
+// Watch executes a long poll for changes to the given key.  handler will be
+// called immediately upon registration to initialize the watch and returns
+// the initial value (as a list, this is what Consul API returns).  There is
+// a chance that Watch can miss updates due to the way the API is implemented.
+// See https://github.com/hashicorp/consul/issues/1761 for details.
+func (w *Watcher) Watch(ctx context.Context, key string, handler WatcherFunc) {
+	w.watching(ctx, key, handler, nil)
+}
+
+// WatchPrefix executes a long poll for changes to anything under the given
+// prefix.  handler will be called immediately upon registration to initialize
+// the watch and returns the initial value (as a list, this is what Consul API
+// returns).  There is a chance that WatchPrefix can miss updates due to the
+// way the API is implemented.  See https://github.com/hashicorp/consul/issues/1761
+// for details.
+func (w *Watcher) WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	q := Query{Param{
+		Name:  "recurse",
+		Value: "",
+	}}
+	w.watching(ctx, prefix, handler, q)
+}
+
+func (w *Watcher) watching(ctx context.Context, key string, handler WatcherFunc, q Query) {
+	if w.MaxAttempts <= 0 {
+		w.MaxAttempts = 10
+	}
+
+	path := "/v1/kv/" + key
+	value := "0"
+
+	attempt := 0
+	for {
+		q = updateParam(q, Param{Name: "index", Value: value})
+		resp := []KeyData{}
+		hdr, err := w.client().do(ctx, "GET", path, q, nil, &resp)
+		if err != nil {
+			attempt += 1
+			if attempt >= w.MaxAttempts {
+				handler(nil, ErrMaxAttempts)
+				return
+			}
+			if v, ok := err.(interface {
+				Temporary() bool
+			}); ok {
+				if v.Temporary() {
+					continue
+				}
+			}
+		} else {
+			attempt = 0
+		}
+		handler(resp, err)
+		value = strconv.FormatUint(hdr.index, 10)
+	}
+}
+
+func updateParam(q Query, new Param) Query {
```

Comment:
How about making this a method of the `Query` type, kind of what `http.Header` does, for example:
```go
q.Add("index", value)
```
```go
func (q *Query) Add(name string, value string) {
    ...
}
```

Reply:
I was thinking that I didn't want to mutate the query passed in, but then I realized I'm passing it by value so I'll do this.

Reply:
No that wasn't it.  It's because Query is a slice type. :-(

Reply:
Is that a problem?

Reply:
Trying to find the proper way to implement it without mutating the underlying array.

My first attempt resulted in the query passed in getting mutated (because cap(q) > len(q) and didn't need a resize).

Reply:
Why is it a problem to mutate the slice? It seems like that's what we want to do here isn't it?

Reply:
Otherwise we can make a copy first:
```go
q = q.Copy()
q.Add("index", value)
```
```go
func (q Query) Copy() Query {
    ...
}
```